### ISSUE TITLE
Possible network tunings to avoid SSL packets fragmentations

### ISSUE THREAD

**Issue by totojack** [emqx/emqx#1534]
#### Environment

- OS: Alpine Linux v3.5 (docker  17.05.0-ce, Centos 7.4.1708)
- Erlang/OTP: R19/8.1
- EMQ: 2.2
--

#### Description

We're experiencing a problem of SSL packets fragmentation with a client that doesn't occur when using TCP listener. I know that this is not an Emqttd related problem, but I'm trying to understand if we can avoid such problems with some tuning on configurations. Certificates are self signed.
These are listeners internal/external configurations (TCP/SSL):

> listener.tcp.external = 0.0.0.0:1883
> listener.tcp.external.acceptors = 64
> listener.tcp.external.max_clients = 1000000
> listener.tcp.external.backlog = 1024
> listener.tcp.external.buffer = 4KB
> listener.tcp.external.nodelay = true
> 
> ## Internal TCP Listener
> listener.tcp.internal = 127.0.0.1:11883
> listener.tcp.internal.acceptors = 16
> listener.tcp.internal.max_clients = 102400
> listener.tcp.internal.backlog = 512
> listener.tcp.internal.tune_buffer = on
> listener.tcp.internal.buffer = 1MB
> listener.tcp.internal.recbuf = 4KB
> listener.tcp.internal.sndbuf = 1MB
> listener.tcp.internal.nodelay = true
> 
> ## External SSL Listener
> listener.ssl.external = 8883
> listener.ssl.external.acceptors = 32
> listener.ssl.external.max_clients = 500000
> listener.ssl.external.access.1 = allow all
> listener.ssl.external.handshake_timeout = 15s
> listener.ssl.external.keyfile = /opt/emqttd/etc/certs/key.pem
> listener.ssl.external.certfile = /opt/emqttd/etc/certs/cert.pem
> listener.ssl.external.cacertfile = /opt/emqttd/etc/certs/cacert.pem
> listener.ssl.external.dhfile = /opt/emqttd/etc/certs/dh-params.pem
> listener.ssl.external.backlog = 1024
> listener.ssl.external.recbuf = 4KB
> listener.ssl.external.sndbuf = 4KB
> listener.ssl.external.buffer = 4KB
> listener.ssl.external.nodelay = true
>

**Comment by totojack** [comment]
Hi.

After googling around the problem, it seems that it depends on ssl fragment size on Erlang ssl implementation (16 KB), without an option to limit this size with a custom value. I've found this (old) post on stackoverflow:

[https://stackoverflow.com/questions/19276598/erlang-ssl-set-max-send-fragment-size
](https://stackoverflow.com/questions/19276598/erlang-ssl-set-max-send-fragment-size
)

We solved it moving ssl termination on an haproxy in front of emqttd (traffic from haproxy to emqttd tcp listener), and limiting the size using this directive into haproxy.cfg:

> tune.ssl.maxrecord 1420

I'm not an Erlang developer, so I don't know if it would be possible to customize this ssl fragment size directly into Emqttd code. Anyway, our situation is quite normal:

- devices communicating using 2G 
- single Emqttd service with SSL port
- need to optimize/limit network traffic in situations of poor bandwith

In your opinion, should Emqttd handle indipendently such cases?

Thanks.
